lib/float/nan.sail

```diff
@@ -0,0 +1,65 @@
+/*==========================================================================*/
+/*     Sail                                                                 */
+/*                                                                          */
+/* Copyright 2024 Intel Corporation                                         */
+/*   Pan Li - pan2.li@intel.com                                             */
+/*                                                                          */
+/* Redistribution and use in source and binary forms, with or without       */
+/* modification, are permitted provided that the following conditions are   */
+/* met:                                                                     */
+/*                                                                          */
+/* 1. Redistributions of source code must retain the above copyright        */
+/*    notice, this list of conditions and the following disclaimer.         */
+/* 2. Redistributions in binary form must reproduce the above copyright     */
+/*    notice, this list of conditions and the following disclaimer in the   */
+/*    documentation and/or other materials provided with the distribution.  */
+/*                                                                          */
+/* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS      */
+/* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT        */
+/* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A  */
+/* PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT       */
+/* HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,   */
+/* SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED */
+/* TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR   */
+/* PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF   */
+/* LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING     */
+/* NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS       */
+/* SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.             */
+/*==========================================================================*/
+
+$ifndef _FLOAT_NAN
+$define _FLOAT_NAN
+
+$include <float/common.sail>
+
+val      float_is_nan : fp_bits -> bool
+function float_is_nan (op) = {
+  let struct {_, exp, mantissa} = float_decompose(op);
+  let is_nan = exp == ones() & mantissa != zeros();
+
+  is_nan
+}
+
+val      float_is_snan : fp_bits -> bool
+function float_is_snan (op) = {
+  let struct {_, exp, mantissa} = float_decompose(op);
+  let highest_bit = sizeof(mantissa) - 1;
```

Comment:
is there any suggestion for get the size(bits) of mantissa here?
For now we will have the below error.
```
Type error:
/home/pli/repos/sail/sail-fork/lib/float/nan.sail:46.20-40:
46 |  let highest_bit = sizeof(mantissa) - 1;
   |                    ^------------------^
   | No overloading for (operator -), tried:
   | * sub_atom
   |    Cannot re-write sizeof(mantissa)
   | * sub_int
   |    Cannot re-write sizeof(mantissa)
```

Meanwhile, `'mantissa` like `let (_, exp, 'mantissa) = float_decompose(op);` won't work here too.

Reply:
Sorry not very familiar as green hand for sail, is there any alternatives?

Reply:
If `mantissa` indicates bits[22 .. 0] (take single as example) of floating-point, the `let highest_bit = mantissa - 1` may not be the value `22` which is the highest bit index of mantissa?
For snsn, we only need to check the highest bits here.

Reply:
Or one silly way is that we can have a helper function to return the highest bit of float, may similar to
```
function float_mantissa_bits (op) = {
  match 'n {
    16 => 10,
    32 => 23,
    64 => 52
  }
}
```

It may works but not that perfect.

Reply:
> Or one silly way is that we can have a helper function to return the highest bit of float, may similar to
> 
> ```
> function float_mantissa_bits (op) = {
>   match 'n {
>     16 => 10,
>     32 => 23,
>     64 => 52
>   }
> }
> ```
> 
> It may works but not that perfect.

Tried this approach but meet another issue for vector bits access. Updated code for aligning the below error, not sure if I missed something.

```
47 |  let tmp_1 = mantissa[highest_bit - 1];
   |              ^-----------------------^
   | No overloading for vector_access, tried:
   | * bitvector_access
   |    No overloading for (operator -), tried:
   |    * sub_atom
   |       Could not unify int('n) and float_bits('ex284#)
   |    * sub_int
   |       Could not unify int('ex174#) and float_bits('ex284#)
   | * plain_vector_access
   |    Could not unify vector('n, 'a) and bitvector((if 'ex284# == 16 then 10 else (if 'ex284# == 32 then 23 else 52)))
   |
   | Caused by /home/pli/repos/sail/sail-fork/lib/float/nan.sail:47.14-39:
   | 47 |  let tmp_1 = mantissa[highest_bit - 1];
   |    |              ^-----------------------^
   |    | Vector access could not be interpreted as a bitfield access
```